src/main/java/com/google/sps/servlets/GenerateServlet.java

```diff
@@ -28,14 +32,143 @@
 
   private static final String FILTER_ARRAY = "clicked-array";
 
+  String[] allPlaces = {"Paris", "New York City", "San Francisco", "London", "Sydney", "Venice"};
+  String[] allDifficulties = {"Easy", "Medium", "Hard"};
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get array of clicked filters and convert to ArrayList<String>
-    Gson gson = new Gson();
     HashSet<String> clickedFilters =
-        gson.fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+        new Gson().fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
```

Comment:
I think using the same request parameter to capture all filter arguments is going to give you some trouble.  From the looks of this code, it seems like the URL request might look like this:
/generate-hunt?clicked-array=[Paris,easy,medium,4,art,kid_friendly]

In this method, you are having to take all your filters, and see which ones pertain to Place vs. Difficulty by using retainAll to see what matches.  That might be a problem if you have something like a number since you won't know what it's for if you have multiple parameters that could be numbers.

But what if the request looked like this:
/generate-hunt?city=Paris&difficulty=easy,medium&num_stops=4&labels=art,kid_friendly

Then you don't need to do retainAll because you already know which ones the user chose.